### PR TITLE
_shard_num fixes

### LINKED ISSUE #12601: GROUP BY over _shard_num doesn't work in distributed tables.
```
SELECT 
    _shard_num,
    count()
FROM remote('127.0.0.{1,2}', system.one)
GROUP BY _shard_num

↙ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) 
Received exception from server (version 20.7.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column _shard_num in block. There are only columns: count(): While executing Remote. 
```

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `GROUP BY`/`LIMIT BY` `_shard_num` (and some caveats for `_shard_num as s` (i.e. `_shard_num` with an alias)/`func(_shard_num)`/`optimize_move_functions_out_of_any`)

Detailed description / Documentation draft:
Let's consider the following example:

Some node got:

```sql
SELECT _shard_num, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY _shard_num
```

The query will be sent to the shard as:

```sql
WITH toUInt32(1) AS _shard_num SELECT _shard_num, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY _shard_num
```

And this creates the some problems (shard is one of the remote nodes):

- GROUP BY on the shard simply removes the `_shard_num` and hence initiator does not receive it to do final aggregation (this has been fixed in affd710743b750131a4c7cb0cffb96a85568ad5b)
- GROUP BY on the shard is rewritten by the `QueryNormalizer`, and in case of `_shard_num` was requested with an alias (`_shard_num AS num`), on the shard `_shard_num` will be alias-to-alias, while on the initiator it will be still parent alias, so the shard, column with name `_shard_num AS foo` will be rewritten to `toUInt32(1) AS foo`, while initiator (since it did not expect from the shard to do projections, i.e. rename columns to aliases) expects `_shard_num` (this has been fixed in 67e05487f45a530a273c0b5c19de7fd3686a2cdc)
- And also in case of `optimize_move_functions_out_of_any=1` (default), `_shard_num` should be not optimized, since in this case shard will have `any(toUInt32(1) AS _shard_num)` while initiator expects `_shard_num`
- `LIMIT BY` expressions will not find column

Fixes: #12601

HEAD:

<details>

- 87199a02bd99813eb26b35b1e19f07c69be550b8
- 367c4ecfae310c1b3036eb65e9fb3186cc19114d
- caf5f98db2ae39dd911f7d1fc0a7c2cc382b1c53

</details>

### CODE REVIEW

**Comment by azat** [comment]
>Functional stateless tests (ubsan) — fail: 1, passed: 2439, skipped: 9

```
2020-11-14 02:26:54 01006_simpod_empty_part_single_column_write:                            [ FAIL ] 1.42 sec. - having stderror:
2020-11-14 02:26:54 [f187b3e7e1e9] 2020.11.14 02:26:54.368113 [ 6927 ] {d2efb991-4e77-4264-b56d-335057dd6509} <Error> executeQuery: Code: 341, e.displayText() = DB::Exception: Exception happened during execution of mutation 'mutation_4.txt' with part '1_1_1_0_3' reason: 'Code: 236, e.displayText() = DB::Exception: Cancelled mutating parts (version 20.12.1.5183)'. This error maybe retryable or not. In case of unretryable error, mutation can be killed with KILL MUTATION query (version 20.12.1.5183) (from [::1]:32884) (in query: ALTER TABLE table_with_empty_part MODIFY COLUMN value Nullable(UInt64)), Stack trace (when copying this message, always include the lines below):
```

>AST fuzzer — Fuzzer failed (1). See the logs

```
Read too much text and still can't parse a query. Aborting.
```

**Comment by azat** [comment]
@alexey-milovidov can you please take a look?

**Comment by azat** [comment]
>Conflicting files
>src/Interpreters/InterpreterSelectWithUnionQuery.h

Going to resolve them after the review (to avoid doing it constantly)

**Comment by akuzm** [comment]
Is this problem related to the fact that we match remote columns by name? This is not the first one, we probably should match them by index, like for `UNION`.

**Comment by azat** [comment]
>Is this problem related to the fact that we match remote columns by name? This is not the first one, we probably should match them by index, like for UNION.

@akuzm Not really, let's consider the following example:

Some node got:

```sql
SELECT _shard_num, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY _shard_num
```

The query will be sent to the shard as:

```sql
WITH toUInt32(1) AS _shard_num SELECT _shard_num, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY _shard_num
```

And this creates the some problems (shard is one of the remote nodes):

- GROUP BY on the shard simply removes the `_shard_num` and hence initiator does not receive it to do final aggregation (this has been fixed in affd710743b750131a4c7cb0cffb96a85568ad5b)
- GROUP BY on the shard is rewritten by the `QueryNormalizer`, and in case of `_shard_num` was requested with an alias (`_shard_num AS num`), on the shard `_shard_num` will be alias-to-alias, while on the initiator it will be still parent alias, so the shard, column with name `_shard_num AS foo` will be rewritten to `toUInt32(1) AS foo`, while initiator (since it did not expect from the shard to do projections, i.e. rename columns to aliases) expects `_shard_num` (this has been fixed in 67e05487f45a530a273c0b5c19de7fd3686a2cdc)
- And also in case of `optimize_move_functions_out_of_any=1` (default), `_shard_num` should be not optimized, since in this case shard will have `any(toUInt32(1) AS _shard_num)` while initiator expects `_shard_num`

**Comment by akuzm** [comment]
> * GROUP BY on the shard is rewritten by the `QueryNormalizer`, and in case of `_shard_num` was requested with an alias (`_shard_num AS num`), on the shard `_shard_num` will be alias-to-alias, while on the initiator it will be still parent alias, so the shard, column with name `_shard_num AS foo` will be rewritten to `toUInt32(1) AS foo`, while initiator (since it did not expect from the shard to do projections, i.e. rename columns to aliases) expects `_shard_num` (this has been fixed in [67e0548](https://github.com/ClickHouse/ClickHouse/commit/67e05487f45a530a273c0b5c19de7fd3686a2cdc))
> * And also in case of `optimize_move_functions_out_of_any=1` (default), `_shard_num` should be not optimized, since in this case shard will have `any(toUInt32(1) AS _shard_num)` while initiator expects `_shard_num`

In these two cases, why does the initiator care which alias the column has on a shard? If the number and the order of resulting columns is the same, the initiator could probably use them regardless of the names.

**Comment by azat** [comment]
>In these two cases, why does the initiator care which alias the column has on a shard?

Using only positions for queries over network does not looks reliable to me, for example there was the following issue - https://github.com/ClickHouse/ClickHouse/pull/10621

**Comment by akuzm** [comment]
> > In these two cases, why does the initiator care which alias the column has on a shard?
> 
> Using only positions in the block for blocks from other storages (via network) does not looks reliable to me, for example there was the following issue - #10621
> 
> For example in the future clickhouse can do some "optimizations" that will change the order (very unlikely, but still) and this will stop working with those nodes

Looks like the names are no more reliable. We are required by standard to use positions, not names, for `UNION`, so maybe using the same rule for remote queries could be simpler. Anyway, just wanted to hear what you think, thanks :)

**Comment by azat** [comment]
Just a friendly reminder that this one still waiting for a review

**Comment by azat** [comment]
@alexey-milovidov can you please take a look?

Also don't miss @akuzm suggestion of matching columns by index for the remote queries (https://github.com/ClickHouse/ClickHouse/pull/16947#issuecomment-735808270), I'm afraid (https://github.com/ClickHouse/ClickHouse/pull/16947#issuecomment-736839876) that it will be less robust and in case of errors it will fail only in case of type mismatach, but this will allow removing two patches from this patches

**Comment by azat** [comment]
@alexey-milovidov Maybe I can make the patch series easier for review, i.e. split them into smaller PRs?

**Comment by azat** [comment]
Any chance that someone will pick this in a the nearest future?

**Comment by alexey-milovidov** [comment]
@azat I did not take this PR for review.

**Comment by alexey-milovidov** [comment]
Let's merge with master again.

**Comment by azat** [comment]
>Fast test — fail: 1, passed: 2320, skipped: 278. Other tests won't be run before fast test will be fixed.

Now JOIN part is broken with #15162 (and is blocked by #8802), so I've replaced the test cases to expect 403 error, but patches are still relevant, since it fixes GROUP BY (and not related to JOIN actually, just because of newly added test cases it failed)

**Comment by azat** [comment]
Tests w/o JOIN passed

**Comment by azat** [comment]
>Yandex synchronization check (only for Yandex employees)
>Yandex third-party checks (only for Yandex employees)

I guess this is due to interface changes...

**Comment by alexey-milovidov** [comment]
Maybe @akuzm can take it for review?

**Comment by akuzm** [comment]
> Also don't miss @akuzm suggestion of matching columns by index for the remote queries ([#16947 (comment)](https://github.com/ClickHouse/ClickHouse/pull/16947#issuecomment-735808270)), I'm afraid ([#16947 (comment)](https://github.com/ClickHouse/ClickHouse/pull/16947#issuecomment-736839876)) that it will be less robust and in case of errors it will fail only in case of type mismatach, but this will allow removing two patches from this patches

@KochetovNicolai  what do you think about this? I think we should be moving away from using the column names for matching e.g. remote inputs, and switch to position-based matching like we do for `UNION`. This patch introduces more code for name-based matching for remote stream. Azat thinks it's more robust, but to me it looks like the opposite -- why can't we require a particular column order from remote stream and instead receive columns in some random order and then search for the ones we need?

File: src/Interpreters/ActionsDAG.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ ActionsDAG::ActionsDAG(const NamesAndTypesList & inputs_)
         addInput(input.name, input.type, true);
 }
 
-ActionsDAG::ActionsDAG(const ColumnsWithTypeAndName & inputs_)
+ActionsDAG::ActionsDAG(const ColumnsWithTypeAndName & inputs_, QueryProcessingStage::Enum to_stage_)
```
Comment (KochetovNicolai) [inline src/Interpreters/ActionsDAG.cpp:37]: We definitely do not want to have `QueryProcessingStage` in ActionsDAG
At least, it can me special flag.

**Comment by azat** [comment]
I'm going to split this PR into separate patches to make it easier to review (with a separate reproducers/tests), the first one is (should be the most obvious) - #23783

File: src/Interpreters/ActionsDAG.cpp (hunk 1 above)
Comment (azat) [reply src/Interpreters/ActionsDAG.cpp:37]: Got it, will take a closer look, thanks!

**Comment by azat** [comment]
Superseded by #24252

**Pull request closed without merging**